Remove mesh from input file and rework input file read in mechanism

davidrudlstorfer opened the pull request:
This PR now removes the mesh entirely from the input file.

Additionally this PR introduces our new API to read in an existing input file as discussed. In my opinion it makes sense that this function is not a free function - but we overwrite the existing function from fourcipp (hopefully this is soon included).

With this change the next step would be to introduce fourcipp into MeshPy. This should be rather straight forward.

davidrudlstorfer commented:
Nice that we check older python versions as well - turns out `Self` is only available from 3.11 onwards

isteinbrecher reviewed (commented):
Thanks a lot for the changes @davidrudlstorfer . I think we can merge this soon, I left some small comments.

The only thing I would like to discuss is regarding the interface and name of `InputFile.from_4C_yaml`: This function returns an input file and a mesh, thus I am not 100% happy that this is inside of InputFile, I see this more as a free function. Also, there is one aspect we (and especially me) didn't think about yet, in 99% of the cases we don't read an existing input file but we convert a CubitPy object to a MeshPy mesh and input file. Thus I am also not sure we want to name this `from_4C_yaml`.

Maybe we can have something like (naming is just a suggestion):
```python
def load_existing_model(yaml_path=None, cubit=None, full_import=False):
    """yaml_path and cubit are mutually exclusive"""
    
    if yaml_path:
        input_file = InputFile.from_4C_yaml(yaml_path)
    else:
        input_file = InputFile()
        input_file.add(cubit.get_dict())
    
    if full_import:
        mesh = input_file.pop_mesh() # This also deletes the mesh sections in input_file
    else:
        mesh = Mesh()

    return input_file, mesh
```

I am curious about your opinion.

davidrudlstorfer commented on src/meshpy/four_c/input_file.py:
```diff
@@ -329,7 +338,7 @@ def _get_section_items(section_name):
-        self.add(mesh)
+        return mesh
 
     def add(self, *args, **kwargs):
```
This makes total sense and it's more or less exactly what I've done in all the tests.

Unfortunately, I do not really understand what "TODO" I should add? To rework this function? This will be done once fourcipp is introduced in the next step

isteinbrecher replied:
Thanks, that is exactly what I was referring to.

davidrudlstorfer commented on src/meshpy/four_c/input_file.py:
```diff
@@ -442,8 +452,9 @@ def write_input_file(
+        mesh: _Mesh,
         *,
         add_header_information: bool = True,
         check_nox: bool = True,
```
Your're 100% right - I actually had the changes prepared to add these only during writing the file. To keep the changes smaller within this PR I reverted them. I will add it back in

davidrudlstorfer replied:
Done

davidrudlstorfer commented on tests/test_meshpy.py:
```diff
@@ -1170,73 +1167,74 @@ def test_meshpy_replace_nodes_geometry_set(
 
 
-@pytest.mark.parametrize("test_type", [None, "full"])
+# TODO change additional identifier
```
I know that I need to improve my ToDo's :D

Across the tests we have like 5 different approaches on how to run these tests where we do a "full" import or only an input file import. These approaches should be standardized

Once we use 

```
@pytest.mark.parametrize(
    ("full_import", "additional_identifier"),
    [(False, "dict_import"), (True, "full_import")],
)
```

another time

```
@pytest.mark.parametrize("full_import", [False, True])
```

as well as 

```
@pytest.mark.parametrize(
    ("full_import", "additional_identifier"),
    [(False, None), (True, "full")],
)
```

Maybe it makes sense to provide a fixture for that

davidrudlstorfer replied:
And similarly - the reference files are also named differently for each case. This was very confusing

davidrudlstorfer replied:
Added a better comment

isteinbrecher replied:
Yes that totally makes sense, thanks for adding the TODOs (and sorry for introducing so many different ways to do this parametrise).

davidrudlstorfer commented on src/meshpy/four_c/input_file.py:
```diff
@@ -216,30 +215,40 @@ def __init__(self, *, yaml_file: _Optional[_Path] = None, cubit=None):
-            file_path:
-                A file path to an existing input file that will be read into this
-                object.
+            input_file_path: A file path to an existing input file that will be read
+                into this object.
+            convert_input_to_mesh: If True, the input file will be converted to a
+                MeshPy mesh.
+
+        Returns:
+            A tuple with the input file and the mesh. If convert_input_to_mesh is
+            False, the mesh will be empty.
```
Done

davidrudlstorfer commented:
@isteinbrecher thanks for your review - I've added a few commits to address your comments

1. I also was very divided if this function makes sense as a class function of input file. My motive for putting it into the InputFile was that we overwrite the fourcipp function in the future. But after giving it some thought the original base function still works and this makes more sense.

I've now created a new file called `model_importer.py` which contains the new `import_4C_model` function. This way the future refactoring is more contained and it also is not directly connected to the 4C input file. You also noted that most of the times a cubit input is loaded. In my opinion it would make sense to also provide a `import_cubit_model` function. This way the functionality is better separated as these two do not have anything in common. But this can be discussed in the future.

I've now also extracted the input file mappings into a separate file as proposed and already discussed in #330

2. I've reworked and already simplified the add function

3. I've again extracted the add_header_information flag to the input file writing mechanism which makes a lot more sense.

4. I've improved the comment regarding the additional identifiers in testing

This would be ready for review again and I am curious about your opinion now :D

davidrudlstorfer commented:
I now also fixed the 3 4C tests which were failing. This was really random because originally something wrong was compared. I needed to adjust the reference file with a definitely missing key. Otherwise the test (correctly) failed.

So prior to this the tests were not working at all.

isteinbrecher commented on src/meshpy/four_c/model_importer.py:
```diff
@@ -0,0 +1,221 @@
+    """Convert an existing input file to a MeshPy mesh with mesh items, e.g.,
+    nodes, elements, element sets, node sets, boundary conditions, materials.
+
+    Note: In the current implementation we cannibalize the mesh sections in
+        the self.sections dictionary. This should be reconsidered and be done
+        in a better way when this function is generalized.
```
This comment is outdated and can be removed (since we can now copy the `input_file` before passing to this function, we can already control this behaviour).

davidrudlstorfer replied:
Removed it

isteinbrecher commented on src/meshpy/four_c/model_importer.py:
```diff
@@ -0,0 +1,221 @@
+) -> _Dict[int, _GeometrySetNodes]:
+    """Add sets of points, lines, surfaces or volumes to the object."""
+
+    # Create the individual geometry sets. The nodes are still integers at this
+    # point. They have to be converted to links to the actual nodes later on.
```
This comment is outdated and can be removed.

davidrudlstorfer replied:
Removed

isteinbrecher commented on src/meshpy/four_c/model_importer.py:
```diff
@@ -0,0 +1,221 @@
+from meshpy.utils.environment import fourcipp_is_available as _fourcipp_is_available
+
+
+def import_4C_model(
```
I like this, also having this in this file. What about the name? Currently we name python functionally related to 4C like: `run_four_c`. Should we name this `import_four_c_model` to be consistent? Instinctively, I don't like the `4C` in the name (numerals and upper case).

I understand your point about having two functions for CubitPy and YAML input and thinking about it, let's do it in your proposed way. We can always merge the functionality if we feel we need to.

davidrudlstorfer replied:
Good catch! Makes sense, renamed it to `import_four_c_model`

isteinbrecher commented on src/meshpy/four_c/model_importer.py:
```diff
@@ -0,0 +1,221 @@
+        return input_file, _Mesh()
+
+
+def boundary_condition_from_dict(
```
Should we mark this (and the following) function private?

davidrudlstorfer replied:
Good idea - changed that

isteinbrecher commented on src/meshpy/four_c/model_importer.py:
```diff
@@ -0,0 +1,221 @@
+    return geometry_sets_in_this_section
+
+
+def sections_to_mesh(input_file: _InputFile) -> _Tuple[_InputFile, _Mesh]:
```
Suggestion: `pop_mesh_sections_to_mesh` or `extract_mesh_sections`, since his is a quite impactful function, I would like to indicate this in the name - but maybe I am overthinking this.

davidrudlstorfer replied:
I've now renamed this to `_extract_mesh_sections`

I'll add an additional comment regarding all of this at the bottom

isteinbrecher commented on src/meshpy/four_c/input_file_mappings.py:
Nice to have this in a file, this improves the readability from `InputFile` a lot!

isteinbrecher commented on src/meshpy/four_c/input_file.py:
```diff
@@ -60,32 +52,7 @@
-
-
-def _get_geometry_set_indices_from_section(
+def get_geometry_set_indices_from_section(
```
This function seems a little out of place there. Suggestion: we could move this to `input_file_utils.py`.

isteinbrecher commented on src/meshpy/four_c/input_file.py:
```diff
@@ -384,41 +152,51 @@ def write_input_file(
         file_path: _Path,
         *,
-        nox_xml_file: _Optional[str] = None,
+        nox_xml_file: bool | str = False,
```
Nice that this type hint works - should we use `Type | None` in favour of `_Optional` in the future?

davidrudlstorfer replied:
I think we should stay with the Optional as it is a bit clearer in my opinion.

Could we convert this option to something using Optional at all? Because Optional automatically implies that it can be None but in this case we have no None?

davidrudlstorfer replied:
Oh that was wrong

We could use `str | None`

I still prefer the Optional

isteinbrecher commented on src/meshpy/four_c/input_file.py:
```diff
@@ -384,41 +152,51 @@ def write_input_file(
+                xml_file_name = "NOT_DEFINED"
+            elif nox_xml_file is True:
+                xml_file_name = _os.path.splitext(file_path)[0] + ".xml"
+            elif isinstance(nox_xml_file, str):
```
Does an final `else` which throws an error make sense?

davidrudlstorfer replied:
Good idea. I've added a simple TypeError

isteinbrecher commented on src/meshpy/four_c/input_file.py:
```diff
@@ -657,32 +403,30 @@ def get_number_of_coupling_conditions(key):
-
-        return yaml_dict
+        _dump_mesh_items(self.sections, "NODE COORDS", mesh.nodes)
+        _dump_mesh_items(self.sections, "STRUCTURE ELEMENTS", mesh.elements)
```
Could you add one TODO here:
```python
# TODO: reset all links and counters set in this method.
```

davidrudlstorfer replied:
Added

isteinbrecher reviewed (commented):
Thanks for the changes @davidrudlstorfer , I really like the way this looks now and think this is a good way to go to in the future!

> So prior to this the tests were not working at all.

Thanks a lot for fixing the tests! Do you know why the issue was not detected previously and the pipeline passed?

davidrudlstorfer commented:
And regarding the failing tests - I did not take the time to look deeper into it why it did not work before. I suspect that somewhere the wrong mesh was utilized, and, therefore, the wrong simulation/data was used.

isteinbrecher reviewed (approved):
Thanks for going the extra mile and tackling all the points @davidrudlstorfer (I actually feel sorry about a few of them, because a lot of the points were originally my mistake...).

Let's this in the current state!

> With the upcoming change I want to simplify the mesh->dict conversion and the section->mesh conversion a lot.
> In a next PR I would clean up a few things across the changed files of this PR. I just want to split it up to keep the better overview.

Cleaning up some small things makes sense! Regarding the next "larger" PRs, I would suggest to introduce fourcipp as a dependency - with simply adapting the testing comparison and removing all legacy string. This should be a relatively nice to review PR. After that, let's have a look at the import and export functionality.

One nice aspect about splitting this up in a few PRs is that I think (or hope) that the upcoming changes will require no changes to the testing (except the comparison with fourcipp).

isteinbrecher commented:
After this is merged, I will try to have a look at all the remaining 4C stuff in `core`, so we know how to handle that before we rework the import and export stuff.
